# Puzzle 2
--- Día 2: I Was Told There Would Be No Math ---


Los elfos se están quedando sin papel de regalo, por lo que deben enviar un pedido para obtener más. Tienen una lista de las dimensiones (largo l, ancho w y alto h) de cada regalo, y solo quieren pedir exactamente lo que necesitan.

Afortunadamente, cada regalo es una caja (un prisma rectangular recto perfecto), lo que facilita un poco el cálculo del papel de envolver requerido para cada regalo: encuentre el área de la superficie de la caja, que es 2 * l * w + 2 * w * h + 2 * h * l. Los elfos también necesitan un poco de papel extra para cada regalo: el área del lado más pequeño.

Por ejemplo:

Un regalo con dimensiones 2x3x4 requiere 2 * 6 + 2 * 12 + 2 * 8 = 52 pies cuadrados de papel de envolver más 6 pies cuadrados de holgura, para un total de 58 pies cuadrados.
Un regalo con dimensiones 1x1x10 requiere 2 * 1 + 2 * 10 + 2 * 10 = 42 pies cuadrados de papel de regalo más 1 pie cuadrado de holgura, para un total de 43 pies cuadrados.
Todos los números de la lista de elfos están en pies. ¿Cuántos pies cuadrados totales de papel de regalo deben pedir?

## Logica de Resolucion

Lo que intentamos buscar es la cantidad total de papel que necesitan los elfos para envolver su regalos, para eso:

* Vamos a guardar el contenido de nuestra entrada en un arhico y luego lo leeremos con las instrucciones `open('input_puzzle', 'r')` y pasaremos el contenido a una variable llamada `mensaje` con la instruccion `.readlines()`.

* Luego usaremos una variable llamada `papel_total` que ira contando la cantidad necesaria de papel para los elfos.

* Recorreremos el mensaje utilizando un ciclo for y como cada linea viene '20x3x11\n', utilizaremos, `.strip()` para poder sacarle los espacio y luego `.split('x')` para que nos devuelva los 3 valores que necesitamos y los almacenaremos en la variable `valores`

* Separaremos los valores en `l` (largo) `w` (ancho) 'h' (alto)

* Utilizaremos la funcion matematica para calcular el `papel_caja`

* Luego crearemos una lista donde almacenaremos las dimensiones, para poder encontrar los 2 valores mas chicos, una vez que encontremos el primer valor, lo sacaremos de la lista y encontraremos el segundo mas chico

* Con `menor1` y `menor2` calculo lo que necesito de `papel_extra`

* En cada iteracion del ciclo voy incrementando el valor de `papel_total` con la suma de `papel_caja + papel_extra`

* Por ultimo mostramos `papel_total` y obtenemos el resultado

## Mis Datos

Los datos que me dieron para el ejercicio si quieres practicar lo puedes encontrar en el archivo input_puzzle2.txt

## Codigo

In [49]:
contenido = open('input_puzzle2.txt', 'r')
mensaje = contenido.readlines()
papel_total = 0
for linea in mensaje:
    valores = linea.strip().split('x')
    l = int(valores[0])
    w = int(valores[1])
    h = int(valores[2])
    papel_caja = 2 * l * w + 2 * w * h + 2 * h * l
    dimensiones = [l, w, h]
    menor1 = min(dimensiones)
    if dimensiones[0] == menor1:
        dimensiones.remove(l)
    elif dimensiones[1] == menor1:
        dimensiones.remove(w)
    else:
        dimensiones.remove(h)
    menor2 = min(dimensiones)
    papel_extra = menor1 * menor2
    papel_total += papel_caja + papel_extra

print(papel_total)

1606483


## Alternativa

In [52]:
contenido = open('input_puzzle2.txt', 'r')
mensaje = contenido.readlines()
papel_total = 0
for linea in mensaje:
    dimensiones = linea.strip().split('x')
    dimensiones = list(map(int, dimensiones))
    l = int(dimensiones[0])
    w = int(dimensiones[1])
    h = int(dimensiones[2])
    papel_caja = 2 * l * w + 2 * w * h + 2 * h * l
    menor1 = min(dimensiones)
    if dimensiones[0] == menor1:
        dimensiones.pop(0)
    elif dimensiones[1] == menor1:
        dimensiones.pop(1)
    else:
        dimensiones.pop(2)
    menor2 = min(dimensiones)
    papel_extra = menor1 * menor2
    papel_total += papel_caja + papel_extra

print(papel_total)

1606483


## Respuesta Parte1: 1606483

## Parte 2

--- La segunda parte ---


Los elfos también se están quedando sin cinta. La cinta tiene el mismo ancho, por lo que solo tienen que preocuparse por la longitud que necesitan para pedir, que nuevamente les gustaría ser exacta.

La cinta necesaria para envolver un regalo es la distancia más corta alrededor de sus lados, o el perímetro más pequeño de cualquier cara. Cada regalo también requiere un lazo hecho con cinta; los pies de cinta necesarios para el arco perfecto es igual a los pies cúbicos de volumen del presente. Sin embargo, no preguntes cómo atan el moño; nunca lo dirán.

Por ejemplo:

Un regalo con dimensiones 2x3x4 requiere 2 + 2 + 3 + 3 = 10 pies de cinta para envolver el presente más 2 * 3 * 4 = 24 pies de cinta para el arco, para un total de 34 pies.
Un regalo con dimensiones 1x1x10 requiere 1 + 1 + 1 + 1 = 4 pies de cinta para envolver el presente más 1 * 1 * 10 = 10 pies de cinta para el arco, para un total de 14 pies.
¿Cuántos pies en total de cinta deberían pedir?

## Logica de Resolucion

En este caso tenemos que estar atentos a los ejemplos y para gregar al ejercicio anterior lo que haremos:

* Plantear la formula de `cinta_caja` como `l * w * h`

* Luego seguiremos necesitando los 2 lados mas chicos y la formula para `cinta_extra` va a quedar como `2 * menor1 + 2 * menor2`

* Luego en cada ciclo iremos incrementando el valor de `cinta_total`

* Por ultimo mostraremos el valor de `cinta_total`

## Codigo

In [50]:
contenido = open('input_puzzle2.txt', 'r')
mensaje = contenido.readlines()
cinta_total = 0
for linea in mensaje:
    valores = linea.strip().split('x')
    l = int(valores[0])
    w = int(valores[1])
    h = int(valores[2])
    cinta_caja = l * h * w
    dimensiones = [l, w, h]
    menor1 = min(dimensiones)
    if dimensiones[0] == menor1:
        dimensiones.remove(l)
    elif dimensiones[1] == menor1:
        dimensiones.remove(w)
    else:
        dimensiones.remove(h)
    menor2 = min(dimensiones)
    cinta_extra = 2 *menor1 +  2 * menor2
    cinta_total += cinta_caja + cinta_extra

print(cinta_total)

3842356


## Alternativa

In [51]:
contenido = open('input_puzzle2.txt', 'r')
mensaje = contenido.readlines()
cinta_total = 0
for linea in mensaje:
    dimensiones = linea.strip().split('x')
    dimensiones = list(map(int, dimensiones))
    l = int(dimensiones[0])
    w = int(dimensiones[1])
    h = int(dimensiones[2])
    cinta_caja = l * w * h
    menor1 = min(dimensiones)
    if dimensiones[0] == menor1:
        dimensiones.pop(0)
    elif dimensiones[1] == menor1:
        dimensiones.pop(1)
    else:
        dimensiones.pop(2)
    menor2 = min(dimensiones)
    cinta_extra = 2 * menor1 + 2 * menor2
    cinta_total += cinta_caja + cinta_extra

print(cinta_total)


3842356


# Respuesta Parte2: 3842356